feat: add desktop shortcuts for Windows applications

enXov opened the pull request:
## Overview
Implements Linux desktop shortcuts (.desktop files) for Windows applications, allowing users to launch apps directly from their desktop environment (GNOME, KDE, etc.).

Closes #121

## What's New
- **Context Menu Integration**: Right-click any app → "Add/Remove Shortcut"
- **Smart Executable Detection**: Auto-detects WinBoat location, prompts user if not found
- **Loading Overlay**: Shows progress when launching apps from shortcuts (Start Container → Wait → Launch)
- **Reactive UI**: Shortcut badges update immediately, persist across navigation
- **Single-Instance Fix**: Second instances exit silently without error dialogs

## Technical Details
- Creates .desktop files in `~/.local/share/applications/`
- Stores app icons in `~/.local/share/winboat/icons/`
- Uses `--launch-app="AppName"` format for IPC communication
- Config persistence for executable path and shortcut list
- Port manager integration for dynamic port handling

## Known Limitations
- Loading overlay uses 3-second static delay (could be made dynamic by detecting RDP window appearance)

## Testing
Tested with:
- Container running and stopped states
- Dev mode (`npm run dev`) and production builds
- KDE launcher integration

---

**Note**: I want everyone who sees this pull request to test these changes. Yes, I tested them and everything works fine for me, but that doesn't mean it will work fine for everyone.

enXov committed:
feat: add desktop shortcuts for Windows applications

Implement Linux .desktop file integration allowing users to launch
Windows apps directly from their desktop environment.

- Add context menu option to create/remove shortcuts
- Auto-detect WinBoat executable location with user fallback
- Display loading overlay during app launch from shortcuts
- Fix single-instance lock to prevent duplicate instance dialogs
- Add reactive shortcut badges in Apps view

Closes #121

svein298 commented:
[Skärmsändning_20251124_220430.webm](https://github.com/user-attachments/assets/aa49db4f-597e-411f-b2a1-ccabfa874cb0)
it do not work on Debian 13

Levev commented on src/main/main.ts:
```diff
@@ -56,18 +56,15 @@ const windowStore = new Store<SchemaType>({
+if (!app.requestSingleInstanceLock()) {
+    // Another instance is already running, exit silently
+    // The second-instance handler in the first instance will receive our command line args
+    app.quit();
```
We should still display the dialog in case WinBoat wasn't invoked with the --launch-app argument

Levev commented on src/main/main.ts:
```diff
@@ -56,18 +56,15 @@ const windowStore = new Store<SchemaType>({
 let mainWindow: BrowserWindow | null = null;
 
-function createWindow() {
-    if (!app.requestSingleInstanceLock()) {
```
If not present, we need to store the realpath of the current argv[0] (or the ARGV0 envvar if it exists) somewhere, so later on we can invoke WinBoat through that path.

WinBoat isn't guaranteed to be present in PATH, so this will solve us a lot of headache later on.

enXov committed:
fix: address code review comments for desktop shortcuts

- Fix single-instance lock to show dialog when not launched with app flags
- Add support for dual launch methods: --launch-app-name and --launch-app-path
- Store WinBoat executable realpath on startup for reliable invocation

Addresses review comments from @Levev:
1. Single instance now only exits silently when invoked with launch flags
2. Implemented --launch-app-name and --launch-app-path for future flexibility
3. Store argv[0] realpath to ensure shortcuts work without PATH dependency

enXov committed:
style: format code with prettier

enXov committed:
Merge branch 'main' into feat/add-shortcuts

enXov committed:
fix: resolve build errors by exporting ContainerStatus and adding GUEST_API_PORT

enXov committed:
refactor: Extract dev wrapper creation and simplify executable path resolution for desktop shortcuts

enXov committed:
fix: run app from shortcuts loading issue

enXov committed:
fix: pgrep check winboat running

enXov commented:
> Works with the fix: [637d68f](https://github.com/TibixDev/winboat/commit/637d68f0307b83433a31bb56f884da0f9239a40f)
> 
> Winboat needs to run from _npm run dev_ though. The shortcut did not work when no winboat or appimage winboat was running

Thanks! is theres any log? should be log the winboat from... [here](https://github.com/TibixDev/winboat/pull/417/commits/eb8399c1c1a9eb72fc6b9bdd163299b7b7cf524a#diff-1590b84ccc4c59f310490f48e7b6b0ee5028d07e6bf9cca4cde41c6d831ae448R146)

Kunstbanause commented:
> Thanks! is theres any log? should be log the winboat from... 

I have a log from the _npm run dev_ build, started in vs code's terminal, but I don't know where the log would be, when it fails to start after using the short-cuts with winboat NOT running beforehand or even the AppImage running. Any idea?

<img width="1073" height="781" alt="image" src="https://github.com/user-attachments/assets/8b415a64-02e2-4179-9c4f-16c8738b258f" />

enXov commented:
> > Thanks! is theres any log? should be log the winboat from...
> 
> I have a log from the _npm run dev_ build, started in vs code's terminal, but I don't know where the log would be, when it fails to start after using the short-cuts with winboat NOT running beforehand or even the AppImage running. Any idea?
> <img alt="image" width="1073" height="781" src="https://private-user-images.githubusercontent.com/3209654/523107222-8b415a64-02e2-4179-9c4f-16c8738b258f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjUwMDk5MDEsIm5iZiI6MTc2NTAwOTYwMSwicGF0aCI6Ii8zMjA5NjU0LzUyMzEwNzIyMi04YjQxNWE2NC0wMmUyLTQxNzktOWM0Zi0xNmM4NzM4YjI1OGYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTIwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEyMDZUMDgyNjQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmFjNjhmNDgwMGFkOTdiYWYwMjQxN2M0ZGY5MTM4MjBiODM5NzFlMjAzYTE2NWZkYTY3YTFkNTBlM2VjYWJmZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.5PQrpS8S5L7BRXSW5P3SrelmIJ0pFUAQqRWP6gUYNxM">

Thanks for the feedback and the screenshot showing the log location (~/.winboat/winboat.log)!

This behavior is by design for development mode. When running via npm run dev, the shortcuts use a wrapper script that requires Winboat to already be running. This is because in dev mode, we can't easily spawn a full npm run dev process from a shortcut - it needs the Vite dev server running.

For production/AppImage users: The shortcuts should work independently - they should start Winboat from the AppImage if it's not running. Could you confirm:

When you created the shortcuts, were you running the AppImage or npm run dev?
Can you check the .desktop file content (in ~/.local/share/applications/) to see what executable path it points to?
If the shortcut was created while running npm run dev, it will point to the dev wrapper (~/.local/bin/winboat-dev-wrapper.sh). You'll need to delete and recreate the shortcuts while running the AppImage for them to work standalone.

The log file at ~/.winboat/winboat.log should help debug production mode issues if the AppImage shortcuts aren't working.

enXov committed:
fix: use AND logic for dev mode detection in desktop shortcuts

Previously, shortcuts would incorrectly use the dev wrapper script when
either app.isPackaged was false OR NODE_ENV was 'development'. This caused
shortcuts to fail in unpacked production builds.

Now both conditions must be true for dev mode, and saved dev wrapper paths
are ignored when running in production mode.

enXov commented:
Sorry about that! My previous explanation was theoretically correct, but there was actually a bug in the code XD

The issue was in the dev mode detection logic - it was using || (OR) instead of && (AND):
```
// Before (buggy):
if (!app.isPackaged || process.env.NODE_ENV === "development")
// After (fixed):
if (!app.isPackaged && process.env.NODE_ENV === "development")
```
This meant shortcuts would incorrectly use the dev wrapper whenever either condition was true, instead of only when both are true.

I've pushed a fix now. Could you please:

Pull the latest changes and rebuild
Delete the existing shortcuts
Recreate them from the production build
They should now work independently without requiring Winboat to already be running. Let me know if it works!

enXov commented:
thansk! somehow on your side try to find program on /tmp/.mount hmmm thanks! lemme quick look now

enXov commented:
okay thanks for your amazing testing, i understand the problem the problem was:

i run the appimage with:

`./winboat-0.9.0-x86_64.AppImage --appimage-extract`
`./squashfs-root/AppRun`

because appimages requires a kind of tool named 'fuse' and i dont want to intsall it and run it from ^ commands, but you probably run the appimage normally

enXov committed:
fix(shortcuts): use APPIMAGE env var for AppImage executable path

When running from an AppImage, the app was using the temporary mount path
(e.g., /tmp/.mount_winboa*/winboat) for desktop shortcuts. This path becomes
invalid after the AppImage is closed, causing shortcuts to fail.

Now the code checks for the APPIMAGE environment variable first, which
contains the actual AppImage file path. This ensures shortcuts work even
after Winboat is closed and reopened.

Fixes issue where shortcuts would show 'Could not find the program' error
on systems running AppImage directly (e.g., Fedora with KDE).

enXov commented:
> > right click a .docx file => open with => winboat(word or whatever) => should be word opened with file in winboat
> 
> You only wrote that as a suggestion, right? Currently that does not work.

Yes i write that as a suggestions, like a summary of what you want. But again its really good future

enXov committed:
fix: container pause

TDS700 commented:
This works great and is a really handy feature to start apps from terminal
Flags like: `--launch-app-name="Notepad"` or using a path seem to have no issues.
Im curious if theres a way to start a whole "Windows Desktop" thing.
In apps tab of winboat gui i can see that windows desktop uses "WINDOWS_DESKTOP" path however `--launch-app-path="WINDOWS_DESKTOP"` doesnt seem to start it and just launches winboat gui. I know its not a regular windows application but if there was a way to launch it from terminal id appreciate it.

enXov committed:
fix: add IPC handler for launch-app-from-shortcut-by-path

This enables launching apps by path from terminal, including internal
apps like WINDOWS_DESKTOP. Previously, --launch-app-path flag was
ignored because there was no IPC handler for it in the renderer.

Fixes launching WINDOWS_DESKTOP from terminal via:
  --launch-app-path="WINDOWS_DESKTOP"

rajibneel commented:
Issue is solved:
https://github.com/TibixDev/winboat/issues/729

<img width="1633" height="773" alt="Screenshot_select-area_20260323135324" src="https://github.com/user-attachments/assets/2a3abfd2-eea1-4f9a-ae16-615151b0feb4" />


<img width="745" height="260" alt="a" src="https://github.com/user-attachments/assets/20c363fe-ded6-4ad5-b0f0-9ec59173cb27" />

The-Nickname commented:
Cool, is there an easy way to add shortcuts via the Winboat UI?

gitneko commented on src/renderer/main.ts:
```diff
@@ -23,3 +29,136 @@ createApp(App)
+        }
+
+        if (!winboat.isOnline.value) {
+            console.error("Container failed to come online in time");
```
It would provide for a better user experience if a "operation timed out" message would be displayed to the user instead of the loading overlay just disappearing. 30 seconds might not be long enough for a "cold start" (container not running yet), so I had this just happen that the overlay disappeared but nothing happened.

enXov replied:
Yes, thats known limits, https://github.com/TibixDev/winboat/pull/417#issue-3553755201, should be fixed before merged
